fix #3569: removing all api references to customResource (#4148)
fabric8io/kubernetes-client: doc/FAQ.md

If you wish to use another HttpClient implementation typically you will exclude kubernetes-httpclient-okhttp and include the other runtime dependency instead.

### What threading concerns are there?

There has been a lot of changes under the covers with thread utilization in the fabric8 client over the 5.x and 6.x releases.  So the exact details of what threads are created / used where will depend on the particular release version.

At the core the thread utilization will depend upon the http client implementation.  Per client OkHttp maintains a pool of threads for task execution.  It will dedicate 2 threads out of that pool per WebSocket connection.  If you have a lot of WebSocket usage (Informer or Watches) with OkHttp, you can expect to see a large number of threads in use - which can potentially exhaust the OkHttp defaults.

With the JDK http client it will only maintain a selector thread and a small worker pool which will be based upon your available processors per client.  It does not matter how many Informers or Watches you run, the same worker pool is shared.

> If it recommended with either client that your Watchers and ResourceEventHandlers do not execute long running tasks.  When using the OkHttp client this inhibits Watch event processing that can timeout the ping for that particular watch.  When using the JDK http client the long running task will inhibit the use of that thread for ALL http processing.  Note that calling other KubernetesClient operations, especially those with waits, can be long-running.  We are working towards providing non-blocking for may of these operations, but until that is available consider using a separate task queue for such work.

Reviewer: ```suggestion
It is recommended with either client that your Watchers and ResourceEventHandlers do not execute long running tasks.  When using the OkHttp client this inhibits Watch event processing that can timeout the ping for that particular watch.  When using the JDK http client the long running task will inhibit the use of that thread for ALL http processing.  Note that calling other KubernetesClient operations, especially those with waits, can be long-running.  We are working towards providing non-blocking mode for many of these operations, but until that is available consider using a separate task queue for such work.
```

If you wish to use another HttpClient implementation typically you will exclude kubernetes-httpclient-okhttp and include the other runtime dependency instead.

### What threading concerns are there?

There has been a lot of changes under the covers with thread utilization in the fabric8 client over the 5.x and 6.x releases.  So the exact details of what threads are created / used where will depend on the particular release version.

At the core the thread utilization will depend upon the http client implementation.  Per client OkHttp maintains a pool of threads for task execution.  It will dedicate 2 threads out of that pool per WebSocket connection.  If you have a lot of WebSocket usage (Informer or Watches) with OkHttp, you can expect to see a large number of threads in use - which can potentially exhaust the OkHttp defaults.

With the JDK http client it will only maintain a selector thread and a small worker pool which will be based upon your available processors per client.  It does not matter how many Informers or Watches you run, the same worker pool is shared.

> It is recommended with either http client that logic you supply via Watchers, ExecListeners, ResourceEventHandlers, Predicates, etc. do not execute long running tasks. 

Reviewer 2: I'd try to highlight this as a callout note box:

```suggestion
> **Note:** It is recommended with either HTTP client implementation that logic you supply via Watchers, ExecListeners, ResourceEventHandlers, Predicates, etc. does not execute long running tasks. 
```

Maybe also suggest the appropriate approach too ---> Schedule those long-running tasks in a separate executor

Author: > Maybe also suggest the appropriate approach too ---> Schedule those long-running tasks in a separate executor

A separate task queue is mentioned below.  We can move or elaborate on that if you want.

Reviewer 2: No, it's fine, I saw the additional explanation later.
The callout is good for diagonal readers, which might stop and pay closer attention to the surrounding paragraphs.

On top of the http client threads the fabric8 client maintains a task thread pool for scheduled tasks and for potentially long-running tasks that are called from WebSocket operations, such as handling input and output streams and ResourceEventHandler call backs.  This thread pool defaults to an unlimited number of cached threads, which will be shutdown when the client is closed - that is a sensible default with either http client as the amount of concurrently running async tasks will typically be low.  If you would rather take full control over the threading use KubernetesClientBuilder.withExecutor or KubernetesClientBuilder.withExecutorSupplier - however note that constraining this thread pool too much will result in a build up of event processing queues.

> Finally the fabric8 client will use 1 thread per PortForward and an additional thread per socket connection - this may be improved upon in the future.

Reviewer 2: 1 thread or 1 extra thread?

I understand that the websocket connection uses a thread, and the WebSocket.Listener implementation uses an extra-thread to process the received messages. right?

Author: It should say instead that "a LocalPortForward uses 1 thread, and every PortForward or connection to the LocalPortForward uses an additional thread."
